Refactored module, removed obsolete files, added tests and CI.

AlexSkrypnyk opened the pull request:
This PR is a major refactor of the module with replacing some procedural approaches with OOP and adding tests.

## Scope
## In scope
- Preserving existing working functionality
- Adding tests
- Fixing coding standards
- Fixing deprecated code
- Support for aliases - will be provided as a separate PR.
- Replacing link paths in documents - needs to be discussed and will be provided as a separate PR.

## Changed
- Refactored core of the module to use OOP approach. Now using `Minisite` class to handle archive and asset management.
- Refactored how asset management works. 
  - Now using `MinisiteAsset` class to work with assets. This class is quite 'thin' now, but will have more functionality once alias support introduced (CRUD records in DB etc).
  - Using Drupal core `ArchiverInterface` and 2 implementation `Zip` and `Tar`. This means that there is a support for `tar` archives.
- Added separate `MinisiteValidator` class to handle specific constraints around archive contents + unit tests.
- Added tests and test scaffolding (will be useful for future tests) to make sure that:
  - Field can be created through UI
  - Field can be set as required
  - Valid archive can be uploaded
  - Invalid archive shows validation error
- Fixed numerous coding standards violations
- Fixed deprecated code
- Removed all code relevant to alias processing as it was not working and was not completely  Drupal 8 compliant. A rewrite of this functionality is coming.
- Removed all code relevant to links processing in the documents. Pending clarifications and discussions.

## How to test
There are 2 ways to test this: 1) within the project itself (same as CI does it) or 2) installing manually into custom or vanilla site.

### Option 1
1. Clone this repo
2. Checkout `8.x-refactor` branch
3. Run `./.circleci/build.sh`. If you have php and sqllite installed, a vanilla Drupal site with module enabled (running on in-built PHP server) will be provisioned.
4. Login to the site (the login link will be available in the output of the script).
5. Perform any manual testing.
Pro tip: run `./.circleci/lint.sh` to lint code and `./.circleci/test.sh` to run all available tests.

### Option 2
1. Download a tarball of `8.x-refactor` branch.
2. Install it into existing site.
3. Perform any manual testing.

AlexSkrypnyk committed:
Updated gitignire to exclude build dir.

AlexSkrypnyk committed:
Updated build sript to allow running on MacOS.

AlexSkrypnyk committed:
Added site installing during the build.

AlexSkrypnyk committed:
Temporary disabled multiple primary keys.

Commit:
Added UI tests.

- Added UI field tests.
- Added config schema.

AlexSkrypnyk committed:
Added roadmap.

Commit:
Update README.md

Commit:
Added required and archive format validation tests. (#3)

Commit:
Refactored module + added tests + removed all non-essential files. (#4)

AlexSkrypnyk commented on .circleci/build.sh:
```diff
@@ -7,19 +7,33 @@ set -e
 composer --working-dir=build require --dev dealerdirect/phpcodesniffer-composer-installer:^0.5
 
-echo "==> Copy module code"
+echo "==> Start inbuilt PHP server in $(pwd)/build/web"
```
this code allows to run the same site build process locally, so it is possible to do development of this module by running `./.circleci/build.sh`

pandaskii replied:
Such a beautiful way

AlexSkrypnyk commented on .circleci/config.yml:
```diff
@@ -15,6 +15,7 @@ job-build: &job-build
       - checkout
       - run: sudo -E apt-get update && sudo -E apt-get install -y libfreetype6-dev libjpeg62-turbo-dev libpng-dev && sudo -E docker-php-ext-install -j$(nproc) iconv && sudo -E docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/ && sudo -E docker-php-ext-install -j$(nproc) gd
       - run: .circleci/build.sh
+      - run: .circleci/lint.sh
```
lint is split into another script to allow running local linting without running tests and the other way around

AlexSkrypnyk commented on .circleci/test.sh:
```diff
@@ -6,11 +6,10 @@ set -e
-
 echo "==> Run tests"
 mkdir -p /tmp/test_results/simpletest
+rm -f /tmp/test.sqlite
```
always use fresh DB for tests

AlexSkrypnyk committed:
Fixed readme.

AlexSkrypnyk commented on composer.json:
```diff
@@ -1,13 +1,13 @@
   ],
   "type": "drupal-module",
-  "license": "Apache-2.0",
+  "license": "GPL-2.0-or-later",
```
correct license name

AlexSkrypnyk commented on minisite.install:
```diff
@@ -5,182 +5,57 @@
   if ($phase == 'runtime') {
-    $requirements['minisite_upload'] = array(
-      'title' => t('Minisite upload directory'),
+    $path = Minisite::getCommonArchiveDir();
```
mainly moved code around to allow paths in Status Report. Like so:
<img width="516" alt="Status_report___Drush_Site-Install" src="https://user-images.githubusercontent.com/378794/69198153-e6451900-0b87-11ea-9b8b-b8fdcd1c0487.png">

AlexSkrypnyk commented on minisite.install:
```diff
@@ -5,182 +5,57 @@
-/**
- * Implements hook_schema().
- */
-function minisite_schema() {
```
we do not need this for now as the schema is required only for alias support

AlexSkrypnyk commented on minisite.module:
```diff
@@ -2,26 +2,15 @@
 /**
  * @file
- * Main functions of the module.
+ * Main functions of the Minisite module.
```
although the file name is the same, the content is completely different.

AlexSkrypnyk commented on minisite.module:
```diff
@@ -2,26 +2,15 @@
-define('MINISITE_MAXFILESIZE', format_size(file_upload_max_size()));
-
-// Load lib for Minisite.
-require_once dirname(__FILE__) . '/includes/minisite.loader.inc';
```
all of these, except for `minisite.theme.in`, were used (partially) for aliases, so they are removed now.

AlexSkrypnyk commented on minisite.module:
```diff
@@ -31,8 +20,35 @@ function minisite_help($route_name, RouteMatchInterface $route_match) {
+ * @return array
+ *   Array of errors, if any.
+ */
+function minisite_validate_archive(FileInterface $file, $content_extensions) {
```
upload validation callback used in widget when the archive is uploaded. Unfortunately, Drupal does not support OOP callbacks in `#upload_validation`, so this has to dangle here

AlexSkrypnyk commented on minisite.theme.inc:
```diff
@@ -5,27 +5,28 @@
  */
 
-use Drupal\Core\Url;
+use Drupal\Core\Link;
```
the theme layer stayed untouched - just code style fixes

AlexSkrypnyk commented on minisite.theme.inc:
```diff
@@ -101,68 +102,17 @@ function template_preprocess_minisite_link(&$variables) {
-/**
- * Implements hook_theme_registry_alter().
- */
-function minisite_theme_registry_alter(&$theme_registry) {
```
this was adding a theme for a minisite page, served through alias. removed for now.

AlexSkrypnyk commented on src/Minisite.php:
```diff
@@ -1,31 +1,214 @@
+  /**
+   * {@inheritdoc}
+   */
+  public static function validateArchive(FileInterface $file, $contentExtensions) {
```
note that this function can be called with file being uploaded (file that has a temporary URI) and file that has already been uploaded (file leaving in the site's `files`).

AlexSkrypnyk commented on src/Minisite.php:
```diff
@@ -1,31 +1,214 @@
+
+    try {
+      /** @var \Drupal\Core\Archiver\ArchiverInterface $archiver */
+      $archiver = \Drupal::getContainer()->get('plugin.manager.minisite_archiver')->getInstance([
```
this is using our own version of ArchiverManager  - see MinisteArchiverManager for more explanation

AlexSkrypnyk commented on src/Plugin/Field/FieldType/MinisiteItem.php:
```diff
@@ -165,15 +160,20 @@ public function fieldSettingsForm(array $form, FormStateInterface $form_state) {
     parent::preSave();
 
-    // Minisite presave.
-    $this->asset_path = Minisite::preSave($this->entity, $this->getEntity());
```
This was one of the 'logic leaking' places. `Minisite` is/was not an entity, so it does not know anything about saving or pre-saving.

Changed this now that a Minisite instance is instantiated and a primary entry point (index file) is retrieved.

AlexSkrypnyk commented on src/Plugin/Field/FieldWidget/MinisiteWidget.php:
```diff
@@ -20,46 +19,27 @@
  * )
  */
 class MinisiteWidget extends FileWidget {
-  /**
```
this was a copy/paste from the parent class

AlexSkrypnyk commented on src/Plugin/Field/FieldWidget/MinisiteWidget.php:
```diff
@@ -20,46 +19,27 @@
   /**
    * {@inheritdoc}
    */
-  public function settingsForm(array $form, FormStateInterface $form_state) {
```
same - not doing anything

AlexSkrypnyk commented on src/Plugin/Field/FieldWidget/MinisiteWidget.php:
```diff
@@ -75,42 +55,26 @@ protected function formMultipleElements(FieldItemListInterface $items, array &$f
   public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
     $element = parent::formElement($items, $delta, $element, $form, $form_state);
 
-    $field_settings = $this->getFieldSettings();
```
the logic here was a bit strange. 

What we want is to check that, if `file_validate_extensions` is one of our existing validators, then it should be using our own extensions to validate rather than defaults.

AlexSkrypnyk commented on src/Plugin/Field/FieldWidget/MinisiteWidget.php:
```diff
@@ -75,42 +55,26 @@ protected function formMultipleElements(FieldItemListInterface $items, array &$f
     $element['#theme'] = 'minisite_widget';
 
-    // Add the additional fields.
-    $element['options']['minisite_alias_status'] = array(
```
removed as related to site alias

AlexSkrypnyk commented on tests/src/Traits/FixtureTrait.php:
```diff
@@ -0,0 +1,195 @@
+use Symfony\Component\Filesystem\Filesystem;
+
+/**
+ * Trait FixtureTrait.
```
this trait help to create fixture archives - we use it to test multiple permutations of the archive. 

This trait has it's own tests.

AlexSkrypnyk commented on tests/src/Unit/MinisiteFixtureTest.php:
```diff
@@ -0,0 +1,131 @@
+/**
+ * Class MinisiteFixtureTest.
+ *
+ * Tests for fixture creation trait (tests for test helpers).
```
tests for tests ❤️

pandaskii commented:
This looks fantastic, thanks for the hard work. 

Two questions, please check the attached screenshots:

1. Allowed number of values - can we force it to 1? 

<img width="727" alt="Screen Shot 2019-11-22 at 11 57 10 am" src="https://user-images.githubusercontent.com/1326673/69389347-dc561e00-0d1f-11ea-9567-f1fa288e6cf3.png">

2. Global blacklisted extensions and allowed file extensions

<img width="689" alt="Screen Shot 2019-11-22 at 11 44 39 am" src="https://user-images.githubusercontent.com/1326673/69389348-dc561e00-0d1f-11ea-9d2c-32a80ea42f0d.png">

AlexSkrypnyk commented:
@pandaskii thank you for reviewing. Your suggested changes are clear - will make the changes + tests.

Commit:
Added support for root-level directories exclusions during validation + tests.

Commit:
Added support for denied extensions + test.

Commit:
Allow to change archive extensions only to user with 'administer site configuration' permission.

Commit:
Locked cardinality to 1 + tests.

Commit:
Enabled Description field and added assets removal on parent entity deletion.

Commit:
Removed obsolete todo.

Commit:
Added aliases support.

pandaskii reviewed (approved):
We are good to a Drupal project release.

thanks for the great work

AlexSkrypnyk commented:
@pandaskii I have another branch with major changes coming to support path aliasing. Adding those changes after release will require significantly more work. I'm actively working on finishing aliasing functionality. Could you please wait for another half-day or so and we will do a release then. Thanks

AlexSkrypnyk commented:
@pandaskii The alias support has now been implemented. We are testing it internally, but it should all work now. Please test yourself and let me know when I can merge this into `8.x` branch.

Once merged, I will create a `8.x-1.0-alpha1` version of the module and will publish it on D.O.

Please note that due to the size of the module, there may still be some issues  - we can address them in `alpha2` etc.

Commit:
Fixed post-delete processing + updated help. (#8)

- Refactored `Minisite` class instance initialisation to no longer save assets.
- Fixed major performance issue with alias event listener firing after core's `RouterListener`, leading to infinite loop when accessing removed assets.
- Fixes deprecated functions to be compatible with core `8.8.0`.
- Updated help text.

Commit:
Updated readme file.

skip ci

Commit:
Updated .gitattributes to not exclude composer.json.

Commit:
Updated CI build scripts and added local dev instructions to readme. (#9)
